src/library/scala/math/BigInt.scala

```diff
@@ -108,14 +109,56 @@ object BigInt {
   /** Implicit conversion from `java.math.BigInteger` to `scala.BigInt`.
    */
   implicit def javaBigInteger2bigInt(x: BigInteger): BigInt = apply(x)
+
+  /** Computes the GCD of the two positive Long arguments, according to the binary GCD algorithm.
+   * Taken from the Spire library. */
+  private def longGcd(_x: Long, _y: Long): Long = {
+    if (_x == 1L || _y == 1L) return 1L
+
+    var x = _x
+    val xz = java.lang.Long.numberOfTrailingZeros(x)
+    x >>= xz
+
+    var y = _y
+    val yz = java.lang.Long.numberOfTrailingZeros(y)
+    y >>= yz
+
+    while (x != y) {
+      if (x > y) {
+        x -= y
+        x >>= java.lang.Long.numberOfTrailingZeros(x)
+      } else {
+        y -= x
+        y >>= java.lang.Long.numberOfTrailingZeros(y)
+      }
+    }
+
+    if (xz < yz) x << xz else x << yz
+  }
+
 }
 
-final class BigInt(val bigInteger: BigInteger)
+/** A type with efficient encoding of arbitrary integers.
+ *
+ * It wraps `java.math.BigInteger`, with optimization for small values that can be encoded in a `Long`.
+ */
+final class BigInt private (private val _bigInteger: BigInteger, private val _long: Long)
   extends ScalaNumber
     with ScalaNumericConversions
     with Serializable
     with Ordered[BigInt]
 {
+  // Class invariant: if the number fits in a Long, then _bigInteger is null and the number is stored in _long
+  //                  otherwise, _bigInteger stores the number and _long = 0L
+  def this(_bigInteger: BigInteger) = this(
+    if (_bigInteger.bitLength <= 63) null else _bigInteger,
+    if (_bigInteger.bitLength <= 63) _bigInteger.longValue else 0L
+  )
+
+  private def this(_long: Long) = this(null, _long)
+
+  def bigInteger: BigInteger = if (_bigInteger eq null) BigInteger.valueOf(_long) else _bigInteger
```

Comment:
We could; though I wanted to limit the overhead of `BigInt` to an additional long. Caching this would mean 1) add another pointer-sized field to BigInt (we already have a `_bigInteger` private field due to the constructor), or 2) reuse the `_bigInteger` field, but in that case a call to `bigInteger` would deoptimize the instance behavior.

WIth this, and other comments, I'd like to get a bit of feedback about the benchmarks that I should use.

Reply:
@johnynek Actually I found a third possibility, see latest commit. Instead of using `_bigInteger eq null` as a tag meaning that the value is in the `_long`, I use `_long == Long.MinValue` as a tag that the value is in the `_bigInteger`.

And in all cases, I can use the `_bigInteger` field, now mutable, to cache the result of the `.bigInteger` call without allocating more memory.